bigflow/dataflow.py

```diff
@@ -1,69 +1,204 @@
+from typing import Dict, List, Union, Optional, Tuple
+
+try:
+    from typing import TypedDict, Literal
+except ImportError:
+    from typing_extensions import TypedDict, Literal
+
+from bigflow.commons import public, build_docker_image_tag
 from bigflow.workflow import Job, JobContext
-from bigflow.workflow import DEFAULT_EXECUTION_TIMEOUT_IN_SECONDS, DEFAULT_PIPELINE_LEVEL_EXECUTION_TIMEOUT_SHIFT_IN_SECONDS
+
+import bigflow.build.reflect
+
 
 logger = logging.getLogger(__file__)
 
 
+class PipelineOptionsDict(TypedDict, total=False):
+    """Standard dataflow pipeline options.
+
+    See sources for full list of supported options.
+    https://github.com/apache/beam/blob/master/sdks/python/apache_beam/options/pipeline_options.py
+    """
+
+    runner: Literal['DirectRunner', 'DataflowRunner']
+    experiments: List[str]
+    streaming: bool
+
+    job_name: str
+    labels: List[str]
+
+    temp_location: str
+    staging_location: str
+
+    project: str
+    region: str
+    zone: str
+    worker_region: str
+    worker_zone: str
+
+    autoscaling_algorithm: Literal['THROUGHPUT_BASED', 'NONE']
+    machine_type: str
+    num_workers: int
+    max_workers: int
+
+    network: str
+    subnetwork: str
+    use_public_ips: bool
+    service_account_email: str
+    worker_disk_type: str
+    disk_size_gb: int
+    flexrs_goal: str
+
+
 @public()
 class BeamJob(Job):
 
+    pipeline_level_execution_timeout_shift = 120  # 2 minutes
+
     def __init__(
             self,
-            id: str,
-            entry_point: typing.Callable[[Pipeline, JobContext, dict], None],
-            pipeline_options: PipelineOptions = None,
+            id: str = None,
+
+            entry_point: typing.Callable[[Pipeline, JobContext, dict], None] = None,
+            entry_point_args: Optional[Tuple] = None,
+            entry_point_kwargs: Optional[Dict] = None,
             entry_point_arguments: typing.Optional[dict] = None,
+
+            pipeline_options: Union[PipelineOptionsDict, PipelineOptions, None] = None,
+
             wait_until_finish: bool = True,
-            execution_timeout_sec: int = DEFAULT_EXECUTION_TIMEOUT_IN_SECONDS,
             test_pipeline: Pipeline = None,
+            execution_timeout_sec: int = None,
+            use_docker_image: typing.Union[str, bool] = False,
+            project_name=None,
+            **kwargs,
     ):
-        if bool(test_pipeline) == bool(pipeline_options):
+        if bool(test_pipeline) == bool(pipeline_options is not None):
             raise ValueError("One of the pipeline and pipeline_options must be provided.")
+
         if not wait_until_finish and execution_timeout_sec:
             raise ValueError("If wait_until_finish_set to False execution_timeout can not be used.")
 
-        self.id = id
-        self.entry_point = entry_point
-        self.entry_point_arguments = entry_point_arguments
-        self.pipeline_options = pipeline_options
+        super().__init__(
```

Comment:
Do we need this? Outside the __init__?

Reply:
Do we need call super `__init__` ? Yes, we need this. But I forgot to add `**kwargs` fixed.